File: examples/ticketdesk/web/src/endurance.ts
import { lstat, readFile, rename, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { performance } from "node:perf_hooks";

import {
  DriverApplicationTransport,
  DriverGeneratedApplicationTransport,
  type DriverApplicationIdentity,
} from "@riffdb/application";

import {
  TicketDeskClient,
  TicketDeskReactiveClient,
  type TicketEventsDelivery,
} from "./generated/client.js";

const TENANTS = ["tenant_alpha", "tenant_beta", "tenant_gamma", "tenant_delta"] as const;
const REQUIRED_COVERAGE = [
  "cold_keys", "events", "hot_keys", "live_queries",
  "multiple_tenants", "reads", "workflows", "writes",
] as const;
type Workload = "events" | "live_queries" | "reads" | "workflows" | "writes";
type Tenant = typeof TENANTS[number];
const LATENCY_BOUNDS_US = [
  50, 100, 200, 400, 800, 1_600, 3_200, 6_400,
  12_800, 25_600, 51_200, 102_400, 204_800, 409_600, 819_200, Number.MAX_SAFE_INTEGER,
] as const;

interface IdentityFile {
  readonly applicationManifestHash: string;
  readonly operationCatalogHash: string;
  readonly contractLineage: string;
  readonly contractVersion: number;
  readonly contractBundleHash: string;
  readonly database: string;
  readonly role: string;
  readonly roleDefinitionHash: string;
  readonly remoteIdentityHash: string;
}

interface ConnectedClient {
  readonly generated: TicketDeskClient;
  readonly reactive: TicketDeskReactiveClient;
  readonly driver: DriverApplicationTransport;
}

class Metrics {
  readonly #path: string;
  readonly #started = Math.floor(Date.now() / 1000);
  #logicalOperations = 0;
  #retainedBytes = 0;
  #eventsEmitted = 0;
  #consumerAcknowledgements = 0;
  #publishing: Promise<void> = Promise.resolve();
  readonly #workloads: Record<Workload, number> = {
    events: 0, live_queries: 0, reads: 0, workflows: 0, writes: 0,
  };
  readonly #tenants: Record<Tenant, number> = {
    tenant_alpha: 0, tenant_beta: 0, tenant_gamma: 0, tenant_delta: 0,
  };
  readonly #latencyCounts = Array.from({ length: LATENCY_BOUNDS_US.length }, () => 0);

  public constructor(path: string) { this.#path = path; }

  public eventEmitted(): void { this.#eventsEmitted += 1; }

  public consumerAcknowledged(): void { this.#consumerAcknowledgements += 1; }

  public async record(workload: Workload, tenant: Tenant, retainedBytes: number, startedAt: number): Promise<void> {
    this.#logicalOperations += 1;
    this.#retainedBytes += retainedBytes;
    this.#workloads[workload] += 1;
    this.#tenants[tenant] += 1;
    const latencyUs = Math.max(0, Math.floor((performance.now() - startedAt) * 1_000));
    const bucket = LATENCY_BOUNDS_US.findIndex((bound) => latencyUs <= bound);
    this.#latencyCounts[bucket < 0 ? this.#latencyCounts.length - 1 : bucket]! += 1;
    if (this.#logicalOperations % 64 === 0) await this.publish();
  }

  public async publish(): Promise<void> {
    const value = `${JSON.stringify({
      schema: "riffdb.alpha-endurance-worker/v1",
      language: "typescript",
      pid: process.pid,
      started_unix_seconds: this.#started,
      logical_operations: this.#logicalOperations,
      transport_attempts: this.#logicalOperations,
      declared_retries: 0,
      error_count: 0,
      events_emitted: this.#eventsEmitted,
      consumer_acknowledgements: this.#consumerAcknowledgements,
      modeled_retained_bytes: this.#retainedBytes,
      latency_bounds_us: LATENCY_BOUNDS_US,
      latency_counts: this.#latencyCounts,
      workloads: this.#workloads,
      tenants: this.#tenants,
    })}\n`;
    const next = `${this.#path}.next`;
    this.#publishing = this.#publishing.then(async () => {
      await writeFile(next, value, { mode: 0o600 });
      await rename(next, this.#path);
    });
    await this.#publishing;
  }
}

const configuration = await requireEnvironment();
const environmentRoot = join(configuration.artifactRoot, "environment-v1");
const metrics = new Metrics(join(environmentRoot, "metrics", "typescript.json"));
await metrics.publish();
const delayMilliseconds = Math.max(1, Math.floor((1_000 * configuration.clients) / configuration.maximumRate));
await Promise.race(TENANTS.map((tenant, index) => runClient(tenant, index, delayMilliseconds)));

async function requireEnvironment(): Promise<{
  readonly artifactRoot: string;
  readonly clients: number;
  readonly maximumRate: number;
  readonly seed: bigint;
}> {
  if (required("RIFFDB_ENDURANCE_LANGUAGE") !== "typescript") throw new Error("TypeScript endurance language differs from the closed manifest");
  const artifactRoot = required("RIFFDB_ENDURANCE_ARTIFACT_ROOT");
  if (!artifactRoot.startsWith("/")) throw new Error("RIFFDB_ENDURANCE_ARTIFACT_ROOT must be absolute");
  const rootStat = await lstat(artifactRoot);
  if (!rootStat.isDirectory() || rootStat.isSymbolicLink()) throw new Error("RIFFDB_ENDURANCE_ARTIFACT_ROOT must be a non-symlink directory");
  const tenants = JSON.parse(required("RIFFDB_ENDURANCE_TENANTS_JSON")) as unknown;
  if (!Array.isArray(tenants) || tenants.length !== TENANTS.length
      || tenants.some((tenant, index) => tenant !== TENANTS[index])) {
    throw new Error("TypeScript endurance tenants differ from the closed manifest");
  }
  const coverage = JSON.parse(required("RIFFDB_ENDURANCE_WORKLOAD_COVERAGE_JSON")) as unknown;
  if (!Array.isArray(coverage) || REQUIRED_COVERAGE.some((name) => !coverage.includes(name))) {
    throw new Error("TypeScript endurance workload coverage is incomplete");
  }
  return {
    artifactRoot,
    clients: boundedNumber("RIFFDB_ENDURANCE_CLIENTS", 4, 4),
    maximumRate: boundedNumber("RIFFDB_ENDURANCE_MAXIMUM_OPERATIONS_PER_SECOND", 1, 1_024),
    seed: BigInt(required("RIFFDB_ENDURANCE_SEED")),
  };
}

function required(name: string): string {
  const value = process.env[name];
  if (value === undefined || value.length < 1 || value.length > 16_384 || value.includes("\0")) {
    throw new Error(`${name} is required`);
  }
  return value;
}

function boundedNumber(name: string, minimum: number, maximum: number): number {
  const value = Number(required(name));
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) throw new Error(`${name} is outside its checked bound`);
  return value;
}

async function connect(role: "seeder" | "application" | "agent"): Promise<ConnectedClient> {
  const wire = JSON.parse(await readFile(join(environmentRoot, `${role}.identity.json`), "utf8")) as IdentityFile;
  const identity: DriverApplicationIdentity = {
    applicationManifestHash: wire.applicationManifestHash,
    operationCatalogHash: wire.operationCatalogHash,
    contractLineage: wire.contractLineage,
    contractVersion: BigInt(wire.contractVersion),
    contractBundleHash: wire.contractBundleHash,
    database: wire.database,
    role: wire.role,
    roleDefinitionHash: wire.roleDefinitionHash,
    remoteIdentityHash: wire.remoteIdentityHash,
  };
  const driver = await DriverApplicationTransport.connect({
    socketPath: join(environmentRoot, `${role}.sock`), identity,
  });
  const transport = new DriverGeneratedApplicationTransport(driver);
  return { generated: new TicketDeskClient(transport, 1), reactive: new TicketDeskReactiveClient(transport), driver };
}

async function runClient(tenant: Tenant, index: number, delayMilliseconds: number): Promise<never> {
  const seeder = await connect("seeder");
  const application = await connect("application");
  const agent = await connect("agent");
  try {
    const namespace = configuration.seed + 2n;
    const organizationId = id(10n, BigInt(index));
    const userId = id(namespace, 100n + BigInt(index));
    const projectId = id(namespace, 200n + BigInt(index));
    const hotTicketId = id(namespace, 300n + BigInt(index));
    await seedClient(seeder.generated, application.generated, tenant, index, organizationId, userId, projectId, hotTicketId);
    const eventParameters = { organization_id: organizationId };
    const eventConsumer = `endurance-typescript-events-${index}`;
    const triageConsumer = `endurance-typescript-triage-${index}`;
    for (let counter = 0; ; counter += 1) {
      let operationStarted = performance.now();
      const slot = counter % 100;
      if (slot < 35) {
        const result = await application.generated.ticketPage({ organization_id: organizationId, ticket_id: hotTicketId });
        if (result.value.outcome !== "Found") throw new Error("TypeScript endurance read lost its hot ticket");
        await metrics.record("reads", tenant, 0, operationStarted);
      } else if (slot < 60) {
        await application.generated.createComment({
          body: `typescript endurance comment ${counter}`, author_id: userId,
          ticket_id: hotTicketId, comment_id: id(namespace, 10_000n + BigInt(index) * 1_000_000n + BigInt(counter)),
          idempotency_key: `endurance-typescript-comment-${index}-${counter}`, organization_id: organizationId,
        });
        await metrics.record("writes", tenant, 512, operationStarted);
      } else if (slot < 70) {
        const batch = await agent.reactive.nextTriageTicket(eventParameters, triageConsumer, 0);
        await metrics.record("workflows", tenant, 0, operationStarted);
        operationStarted = performance.now();
        const item = batch.items[0];
        if (item !== undefined) {
          await agent.reactive.reactComment(eventParameters, triageConsumer, item, {
            body: "typescript contextual endurance reaction", author_id: item.delivery.event.reporter_id,
            ticket_id: item.delivery.event.ticket_id,
            comment_id: id(namespace, 20_000n + BigInt(index) * 1_000_000n + BigInt(counter)),
            idempotency_key: `endurance-typescript-reaction-${index}-${counter}`, organization_id: organizationId,
          });
          metrics.consumerAcknowledged();
          await metrics.record("workflows", tenant, 512, operationStarted);
        }
      } else if (slot < 80) {
        const stream = agent.reactive.ticketEvents(eventParameters, eventConsumer, {
          batchLimit: 1, inFlightLimit: 4, leaseSeconds: 60, maximumWaitMs: 0,
        });
        const iterator = stream[Symbol.asyncIterator]();
        const result = await iterator.next();
        await iterator.return?.();
        if (result.done === true) throw new Error("TypeScript event stream ended before a batch");
        await metrics.record("events", tenant, 0, operationStarted);
        operationStarted = performance.now();
        const delivery: TicketEventsDelivery | undefined = result.value.events[0];
        if (delivery !== undefined) {
          await agent.reactive.ackTicketEvents(eventParameters, eventConsumer, delivery);
          metrics.consumerAcknowledged();
          await metrics.record("events", tenant, 0, operationStarted);
        }
      } else {
        const abort = new AbortController();
        const timer = setTimeout(() => abort.abort(), 5_000);
        try {
          const stream = application.reactive.watchTicketQueueWatch({ organization_id: organizationId, project_id: projectId }, undefined, abort.signal);
          const iterator = stream[Symbol.asyncIterator]();
          const result = await iterator.next();
          await iterator.return?.();
          if (result.done === true) throw new Error("TypeScript live query ended before its snapshot");
          await metrics.record("live_queries", tenant, 0, operationStarted);
        } finally {
          clearTimeout(timer);
        }
      }
      if (slot === 69) {
        operationStarted = performance.now();
        const ordinal = BigInt(Math.floor(counter / 100) % 4_096);
      const created = await application.generated.createTicket({
          title: `TypeScript cold ticket ${index}-${ordinal}`, status: "Open",
          ticket_id: id(namespace, 30_000n + BigInt(index) * 4_096n + ordinal), project_id: projectId,
          assignee_id: userId, reporter_id: userId,
          idempotency_key: `endurance-typescript-cold-${index}-${ordinal}`, organization_id: organizationId,
      });
      if (!created.replayed && created.outcome.outcome === "Created") metrics.eventEmitted();
      await metrics.record("workflows", tenant, 768, operationStarted);
      }
      await new Promise<void>((resolve) => setTimeout(resolve, delayMilliseconds));
    }
  } finally {
    await Promise.all([seeder.driver.shutdown(), application.driver.shutdown(), agent.driver.shutdown()]);
  }
}

async function seedClient(
  seeder: TicketDeskClient, application: TicketDeskClient, tenant: Tenant, index: number,
  organizationId: string, userId: string, projectId: string, ticketId: string,
): Promise<void> {
  let operationStarted = performance.now();
  await seeder.createOrganization({ name: `Endurance ${tenant}`, organization_id: organizationId, idempotency_key: `endurance-organization-${tenant}` });
  await metrics.record("writes", tenant, 512, operationStarted);
  operationStarted = performance.now();
  await seeder.createUser({ email: `typescript-${index}@${tenant}.example.test`, user_id: userId, display_name: `TypeScript endurance ${index}`, idempotency_key: `endurance-typescript-user-${index}`, organization_id: organizationId });
  await metrics.record("writes", tenant, 512, operationStarted);
  operationStarted = performance.now();
  await seeder.createProject({ name: `TypeScript endurance ${index}`, project_id: projectId, idempotency_key: `endurance-typescript-project-${index}`, organization_id: organizationId });
  await metrics.record("writes", tenant, 512, operationStarted);
  operationStarted = performance.now();
  const created = await application.createTicket({ title: `TypeScript hot ticket ${index}`, status: "Open", ticket_id: ticketId, project_id: projectId, assignee_id: userId, reporter_id: userId, idempotency_key: `endurance-typescript-hot-${configuration.seed}-${index}`, organization_id: organizationId });
  if (!created.replayed && created.outcome.outcome === "Created") metrics.eventEmitted();
  await metrics.record("writes", tenant, 768, operationStarted);
  await metrics.publish();
}

function id(namespace: bigint, value: bigint): string {
  const prefix = (namespace & 0xffff_ffffn).toString(16).padStart(8, "0");
  const suffix = (value & 0xffff_ffff_ffffn).toString(16).padStart(12, "0");
  return `${prefix}-0000-8000-8000-${suffix}`;
}

File: examples/ticketdesk/src/bin/endurance.rs
#![forbid(unsafe_code)]

//! Long-lived public-surface TicketDesk workload for the alpha endurance gate.

use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::num::NonZeroU32;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use riffdb_client_rust::{
    AttemptBudget, CallMetadata, DatabaseAlias, EventConsumerOptions, StableApplicationClient,
    load_protected_bearer_credential,
};
use riffdb_config::{
    CanonicalHttpsEndpoint, ProtectedFilePath, TlsClientConfig, TlsServerIdentity,
};
use riffdb_ticketdesk::{
    CreateCommentInput, CreateOrganizationInput, CreateProjectInput, CreateTicketInput,
    CreateUserInput, TicketDeskClient, TicketEventsConsumer, TicketEventsEvent, TicketEventsParams,
    TicketPageParams, TicketPageResult, TicketQueueWatchParams, TriageTicketConsumer,
    TriageTicketParams,
};
use tokio::sync::Mutex;

const TENANTS: [&str; 4] = [
    "tenant_alpha",
    "tenant_beta",
    "tenant_gamma",
    "tenant_delta",
];
const EXPECTED_TENANTS_JSON: &str =
    "[\"tenant_alpha\",\"tenant_beta\",\"tenant_gamma\",\"tenant_delta\"]";
const WORKLOADS: [&str; 5] = ["events", "live_queries", "reads", "workflows", "writes"];

type WorkerResult<T> = Result<T, Box<dyn std::error::Error + Send + Sync>>;

#[derive(Debug)]
struct Metrics {
    started_seconds: u64,
    operations: u64,
    transport_attempts: u64,
    modeled_retained_bytes: u64,
    workloads: BTreeMap<&'static str, u64>,
    tenants: BTreeMap<&'static str, u64>,
}

impl Metrics {
    fn new() -> WorkerResult<Self> {
        let started_seconds = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
        Ok(Self {
            started_seconds,
            operations: 0,
            transport_attempts: 0,
            modeled_retained_bytes: 0,
            workloads: WORKLOADS.into_iter().map(|name| (name, 0)).collect(),
            tenants: TENANTS.into_iter().map(|name| (name, 0)).collect(),
        })
    }

    fn record(&mut self, workload: &'static str, tenant: &'static str, retained_bytes: u64) {
        self.operations = self.operations.saturating_add(1);
        self.transport_attempts = self.transport_attempts.saturating_add(1);
        self.modeled_retained_bytes = self.modeled_retained_bytes.saturating_add(retained_bytes);
        *self.workloads.get_mut(workload).expect("closed workload") += 1;
        *self.tenants.get_mut(tenant).expect("closed tenant") += 1;
    }

    fn json(&self) -> String {
        format!(
            concat!(
                "{{\"schema\":\"riffdb.alpha-endurance-worker/v1\",",
                "\"language\":\"rust\",\"pid\":{},\"started_unix_seconds\":{},",
                "\"logical_operations\":{},\"transport_attempts\":{},",
                "\"declared_retries\":0,\"error_count\":0,",
                "\"modeled_retained_bytes\":{},",
                "\"workloads\":{{\"events\":{},\"live_queries\":{},\"reads\":{},",
                "\"workflows\":{},\"writes\":{}}},",
                "\"tenants\":{{\"tenant_alpha\":{},\"tenant_beta\":{},",
                "\"tenant_gamma\":{},\"tenant_delta\":{}}}}}\n"
            ),
            std::process::id(),
            self.started_seconds,
            self.operations,
            self.transport_attempts,
            self.modeled_retained_bytes,
            self.workloads["events"],
            self.workloads["live_queries"],
            self.workloads["reads"],
            self.workloads["workflows"],
            self.workloads["writes"],
            self.tenants["tenant_alpha"],
            self.tenants["tenant_beta"],
            self.tenants["tenant_gamma"],
            self.tenants["tenant_delta"],
        )
    }
}

struct Clients {
    seeder: TicketDeskClient,
    application: TicketDeskClient,
    agent: TicketDeskClient,
}

#[tokio::main]
async fn main() -> WorkerResult<()> {
    require_environment()?;
    let artifact_root = absolute_environment_path("RIFFDB_ENDURANCE_ARTIFACT_ROOT")?;
    let environment_root = artifact_root.join("environment-v1");
    let metrics_path = environment_root.join("metrics/rust.json");
    let metrics = Arc::new(Mutex::new(Metrics::new()?));
    publish_metrics(&metrics_path, &metrics).await?;

    let clients = required_u64("RIFFDB_ENDURANCE_CLIENTS")?;
    if clients != 4 {
        return Err("Rust endurance worker requires exactly four clients".into());
    }
    let maximum_rate = required_u64("RIFFDB_ENDURANCE_MAXIMUM_OPERATIONS_PER_SECOND")?;
    if !(1..=1_024).contains(&maximum_rate) {
        return Err("Rust endurance rate exceeds its checked bound".into());
    }
    let seed = required_u64("RIFFDB_ENDURANCE_SEED")?;
    let delay = Duration::from_nanos(
        1_000_000_000_u64
            .saturating_mul(clients)
            .div_ceil(maximum_rate),
    );

    let mut tasks = Vec::new();
    for client_index in 0..clients {
        let metrics = Arc::clone(&metrics);
        let metrics_path = metrics_path.clone();
        let environment_root = environment_root.clone();
        tasks.push(tokio::spawn(async move {
            run_client(
                client_index,
                seed,
                delay,
                &environment_root,
                &metrics_path,
                metrics,
            )
            .await
        }));
    }
    for task in tasks {
        task.await??;
    }
    Ok(())
}

async fn run_client(
    client_index: u64,
    seed: u64,
    delay: Duration,
    environment_root: &Path,
    metrics_path: &Path,
    metrics: Arc<Mutex<Metrics>>,
) -> WorkerResult<()> {
    let tenant = TENANTS[usize::try_from(client_index)?];
    let organization_id = id(10, client_index);
    let user_id = id(seed, 100 + client_index);
    let project_id = id(seed, 200 + client_index);
    let hot_ticket_id = id(seed, 300 + client_index);
    let mut clients = Clients {
        seeder: connect(environment_root, "seeder.credential").await?,
        application: connect(environment_root, "application.credential").await?,
        agent: connect(environment_root, "agent.credential").await?,
    };

    seed_client(
        &mut clients,
        tenant,
        &organization_id,
        &user_id,
        &project_id,
        &hot_ticket_id,
        seed,
        client_index,
        &metrics,
    )
    .await?;
    publish_metrics(metrics_path, &metrics).await?;

    let event_consumer = TicketEventsConsumer {
        parameters: TicketEventsParams {
            organization_id: organization_id.clone(),
        },
        consumer_name: format!("endurance-rust-events-{client_index}"),
    };
    let triage_consumer = TriageTicketConsumer {
        parameters: TriageTicketParams {
            organization_id: organization_id.clone(),
        },
        consumer_name: format!("endurance-rust-triage-{client_index}"),
    };
    let mut counter = 0_u64;
    loop {
        let slot = counter % 100;
        if slot < 35 {
            let page = clients
                .application
                .ticket_page(TicketPageParams {
                    organization_id: organization_id.clone(),
                    ticket_id: hot_ticket_id.clone(),
                })
                .await?;
            if !matches!(page, TicketPageResult::Found(_)) {
                return Err("Rust endurance read lost its hot ticket".into());
            }
            record(&metrics, "reads", tenant, 0).await;
        } else if slot < 60 {
            clients
                .application
                .create_comment(CreateCommentInput {
                    body: format!("rust endurance comment {counter}"),
                    author_id: user_id.clone(),
                    ticket_id: hot_ticket_id.clone(),
                    comment_id: id(seed, 10_000 + client_index * 1_000_000 + counter),
                    idempotency_key: format!("endurance-rust-comment-{client_index}-{counter}"),
                    organization_id: organization_id.clone(),
                })
                .await?;
            record(&metrics, "writes", tenant, 512).await;
        } else if slot < 70 {
            let batch = clients
                .agent
                .next_triage_ticket(triage_consumer.clone(), 0)
                .await?;
            record(&metrics, "workflows", tenant, 0).await;
            if let Some(item) = batch.items.first() {
                let TicketEventsEvent::TicketCreated(event) = &item.event;
                clients
                    .agent
                    .react_comment(
                        &triage_consumer,
                        item,
                        &CreateCommentInput {
                            body: "rust contextual endurance reaction".to_owned(),
                            author_id: event.reporter_id.clone(),
                            ticket_id: event.ticket_id.clone(),
                            comment_id: id(seed, 20_000 + client_index * 1_000_000 + counter),
                            idempotency_key: format!(
                                "endurance-rust-reaction-{client_index}-{counter}"
                            ),
                            organization_id: organization_id.clone(),
                        },
                    )
                    .await?;
                record(&metrics, "workflows", tenant, 512).await;
            }
        } else if slot < 80 {
            let batch = clients
                .agent
                .next_ticket_events(
                    event_consumer.clone(),
                    EventConsumerOptions {
                        batch_limit: 1,
                        in_flight_limit: 4,
                        lease_seconds: 60,
                        maximum_wait_nanos: 0,
                    },
                )
                .await?;
            record(&metrics, "events", tenant, 0).await;
            if let Some(delivery) = batch.events.first() {
                clients
                    .agent
                    .ack_ticket_events(&event_consumer, delivery)
                    .await?;
                record(&metrics, "events", tenant, 0).await;
            }
        } else {
            let mut stream = clients
                .application
                .watch_ticket_queue_watch(
                    TicketQueueWatchParams {
                        organization_id: organization_id.clone(),
                        project_id: project_id.clone(),
                    },
                    None,
                )
                .await?;
            let update = tokio::time::timeout(Duration::from_secs(5), stream.message()).await??;
            if update.is_none() {
                return Err("Rust endurance live query ended before its snapshot".into());
            }
            record(&metrics, "live_queries", tenant, 0).await;
        }

        if slot == 69 {
            let cold_ordinal = (counter / 100) % 4_096;
            clients
                .application
                .create_ticket(CreateTicketInput {
                    title: format!("Rust cold ticket {client_index}-{cold_ordinal}"),
                    status: "Open".to_owned(),
                    ticket_id: id(seed, 30_000 + client_index * 4_096 + cold_ordinal),
                    project_id: project_id.clone(),
                    assignee_id: user_id.clone(),
                    reporter_id: user_id.clone(),
                    idempotency_key: format!("endurance-rust-cold-{client_index}-{cold_ordinal}"),
                    organization_id: organization_id.clone(),
                })
                .await?;
            record(&metrics, "workflows", tenant, 768).await;
        }
        counter = counter.saturating_add(1);
        if counter.is_multiple_of(64) {
            publish_metrics(metrics_path, &metrics).await?;
        }
        tokio::time::sleep(delay).await;
    }
}

#[allow(clippy::too_many_arguments)]
async fn seed_client(
    clients: &mut Clients,
    tenant: &'static str,
    organization_id: &str,
    user_id: &str,
    project_id: &str,
    ticket_id: &str,
    seed: u64,
    client_index: u64,
    metrics: &Arc<Mutex<Metrics>>,
) -> WorkerResult<()> {
    clients
        .seeder
        .create_organization(CreateOrganizationInput {
            name: format!("Endurance {tenant}"),
            organization_id: organization_id.to_owned(),
            idempotency_key: format!("endurance-organization-{tenant}"),
        })
        .await?;
    record(metrics, "writes", tenant, 512).await;
    clients
        .seeder
        .create_user(CreateUserInput {
            email: format!("rust-{client_index}@{tenant}.example.test"),
            user_id: user_id.to_owned(),
            display_name: format!("Rust endurance {client_index}"),
            idempotency_key: format!("endurance-rust-user-{client_index}"),
            organization_id: organization_id.to_owned(),
        })
        .await?;
    record(metrics, "writes", tenant, 512).await;
    clients
        .seeder
        .create_project(CreateProjectInput {
            name: format!("Rust endurance {client_index}"),
            project_id: project_id.to_owned(),
            idempotency_key: format!("endurance-rust-project-{client_index}"),
            organization_id: organization_id.to_owned(),
        })
        .await?;
    record(metrics, "writes", tenant, 512).await;
    clients
        .application
        .create_ticket(CreateTicketInput {
            title: format!("Rust hot ticket {client_index}"),
            status: "Open".to_owned(),
            ticket_id: ticket_id.to_owned(),
            project_id: project_id.to_owned(),
            assignee_id: user_id.to_owned(),
            reporter_id: user_id.to_owned(),
            idempotency_key: format!("endurance-rust-hot-{seed}-{client_index}"),
            organization_id: organization_id.to_owned(),
        })
        .await?;
    record(metrics, "writes", tenant, 768).await;
    Ok(())
}

async fn connect(environment_root: &Path, credential_name: &str) -> WorkerResult<TicketDeskClient> {
    let endpoint = required("RIFFDB_ENDURANCE_ENDPOINT")?;
    let tls = TlsClientConfig::new(
        CanonicalHttpsEndpoint::parse(&endpoint)?,
        ProtectedFilePath::new(environment_root.join("ca.pem"))?,
        TlsServerIdentity::parse("127.0.0.1")?,
        Duration::from_secs(5),
        Duration::from_secs(30),
        NonZeroU32::new(4).expect("nonzero pool bound"),
        NonZeroU32::new(64).expect("nonzero stream bound"),
    )?;
    let credential = load_protected_bearer_credential(&environment_root.join(credential_name))?;
    let metadata =
        CallMetadata::authenticated(credential).with_database(DatabaseAlias::default_alias());
    Ok(TicketDeskClient::new(
        StableApplicationClient::connect_verified_tls(&tls).await?,
        metadata,
        AttemptBudget::new(1).expect("positive attempt budget"),
    ))
}

async fn record(
    metrics: &Arc<Mutex<Metrics>>,
    workload: &'static str,
    tenant: &'static str,
    retained_bytes: u64,
) {
    metrics
        .lock()
        .await
        .record(workload, tenant, retained_bytes);
}

async fn publish_metrics(path: &Path, metrics: &Arc<Mutex<Metrics>>) -> WorkerResult<()> {
    let next = path.with_extension("json.next");
    let value = metrics.lock().await.json();
    fs::write(&next, value)?;
    fs::rename(next, path)?;
    Ok(())
}

fn require_environment() -> WorkerResult<()> {
    if required("RIFFDB_ENDURANCE_LANGUAGE")? != "rust"
        || required("RIFFDB_ENDURANCE_TENANTS_JSON")? != EXPECTED_TENANTS_JSON
    {
        return Err("Rust endurance worker environment differs from the closed manifest".into());
    }
    let coverage = required("RIFFDB_ENDURANCE_WORKLOAD_COVERAGE_JSON")?;
    for required_name in [
        "cold_keys",
        "events",
        "hot_keys",
        "live_queries",
        "multiple_tenants",
        "reads",
        "workflows",
        "writes",
    ] {
        if !coverage.contains(&format!("\"{required_name}\"")) {
            return Err("Rust endurance workload coverage is incomplete".into());
        }
    }
    Ok(())
}

fn required(name: &str) -> WorkerResult<String> {
    let value = env::var(name).map_err(|_| format!("{name} is required"))?;
    if value.is_empty() || value.len() > 16_384 || value.contains('\0') {
        return Err(format!("{name} is invalid").into());
    }
    Ok(value)
}

fn required_u64(name: &str) -> WorkerResult<u64> {
    Ok(required(name)?.parse()?)
}

fn absolute_environment_path(name: &str) -> WorkerResult<PathBuf> {
    let path = PathBuf::from(required(name)?);
    if !path.is_absolute() || path.is_symlink() {
        return Err(format!("{name} must be an absolute non-symlink path").into());
    }
    Ok(path)
}

fn id(namespace: u64, value: u64) -> String {
    format!(
        "{namespace:08x}-0000-8000-8000-{value:012x}",
        namespace = namespace & 0xffff_ffff,
        value = value & 0xffff_ffff_ffff,
    )
}
